arq/worker.py

```diff
@@ -357,19 +381,100 @@ async def main(self) -> None:
         if self.on_startup:
             await self.on_startup(self.ctx)
 
-        async for _ in poll(self.poll_delay_s):
-            await self._poll_iteration()
+        await self.create_consumer_group()
+
+        done, pending = await asyncio.wait(
+            [
+                asyncio.ensure_future(self.run_delayed_queue_poller()),
+                asyncio.ensure_future(self.run_stream_reader()),
+                asyncio.ensure_future(self.run_idle_consumer_cleanup()),
+            ],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+
+        for task in pending:
+            task.cancel()
+
+        await asyncio.gather(*pending, return_exceptions=True)
+
+        for task in done:
+            task.result()
+
+    async def run_stream_reader(self) -> None:
+        while True:
+            await self._read_stream_iteration()
```

Comment:
When using this branch in production, we found that this task for some reason didn't cancel properly preventing graceful shutdown on a SIGINT. 

[This is what I ended up doing](https://github.com/hunchdata/arq/commit/9951c0c626ae9ca39a7f85ac1acbb5e2e8953d99), but not sure if there is something more idiomatic